docs/adr/0005-keep-document-history-in-sync-with-rabbit-mq.md

```diff
@@ -0,0 +1,147 @@
+see the events for a particular document, as well as filtering by event type and date. 
+
+An JSON representation of event object will look like this:
+
+```json
+{
+  "id": 115,
+  "action": "HostContentUpdateJob",
+  "user_uid": null,
+  "created_at": "2024-11-28T14:14:11.375Z",
+  "updated_at": "2024-11-28T14:14:11.375Z",
+  "request_id": "91cfbab2f3ff8889ff55a1c7b308d60c",
+  "content_id": "0c643225-b5ae-4bd4-8c5d-9d8911433e28",
+  "payload": {
+    "locale": "en",
+    "message": "Host content updated by content block update",
+    "content_id": "0c643225-b5ae-4bd4-8c5d-9d8911433e28",
+    "source_block": {
+      "title": " Universal Credit Helpline ",
+      "content_id": "a55a917b-740f-466b-9b31-9a9df4526de4",
+    }
+  }
+}
+```
+
+When a document is loaded in Whitehall, we could then call the API and weave these events into the timeline.
+However, this is complicated by the fact that Whitehall's document history is paginated, so we won't necessarily
+have the full Whitehall history at load time and won't necessarily know the full date window of Publishing events
+to fetch. For example:
+
+A document has the following range of event datetimes for the first page:
+
+```
+2024-03-23T09:23:00
+.....
+2023-12-10T11:13:00
+```
+
+And a range of event datetimes for the second page
+
+```
+2023-11-22T12:27:00
+...
+2023-09-12T15:17:00
+```
+
+If we have an event that happens between `2023-11-22T12:27:00` (the newest event for the second page) and 
+`2023-12-10T11:13:00` (the oldest event for the first page) it won't get picked up because it doesn't occur 
+within that range of events.
+
+We could get around this by making a request to get the datetime of the first event on the next page, thus
+giving us a full window of dates to interleave, but this makes an already [complex class][1] harder to understand.
+
+Additionally, making an extra database query and calling out to an API endpoint could have performance impacts.
+
+It's also worth considering that currently, we display 10 events on each "page" of results. If we are interleaving
+new events with each page of results, this could be confusing for the user if they only expect to see 10 results.
+
+Another solution could be sending a request to the Publishing API endpoint before we fetch the history and then creating
+new events, however:
+
+1. This will result in an API call every time a user views a document; and
+2. Carrying out an INSERT query on a GET request isn't a pattern we want to encourage
+
+## Solution 2: Add a new message consumer in Whitehall
+
+This would involve setting up a new RabbitMQ message topic in Publishing API that sends 
+messages when a content block update triggers a change to a document. This would be a brand new 
+topic that contains a thin message that includes the `content_id` of the document that has
+been updated, when it was updated and information about the content block that triggered the update:
```

Comment:
This interesting. In my imagination we were going to send just a message for the content block and then leave Whitehall to do the lookup of all the documents that link to the block. However, I realise now that would mean searching through govspeak because Whitehall doesn't have a structured reference to content blocks. I can see now why the original plan was to use the existing message, because we'd now have to send two messages for each document. Hmm.

Whitehall does a thing where it parses out the references from Govspeak to contacts and to other editions each time an edition is saved ([link](https://github.com/alphagov/whitehall/blob/main/app/services/service_listeners/edition_dependencies_populator.rb)). We could do the same with content block references perhaps, but I have not really thought through the consequences of that (performance etc.). It would allow us to just send the content block ID to Whitehall and update the document history though.

Reply:
We have got a similar thing (contained within the [Content Block Tools Gem](https://github.com/alphagov/govuk_content_block_tools)) that we use to extract references from Govspeak (It's actually [used in Whitehall at the mo](https://github.com/alphagov/whitehall/blob/50acb8baa46c0fe3da0872486421ad53bc6b922e/app/helpers/admin/admin_govspeak_helper.rb) too), so that's entirely a thing we could do